[CBRD-23490] create online index (#60)
CUBRID/cubrid-manual: en/sql/schema/index_stmt.rst

    
    *   The leap second support parameter (:ref:`timezone-parameters`) should not be changed if database contains indexes or function index on columns of type TIMESTAMP or TIMESTAMP WITH LOCAL TIME ZONE. 
    
    *   The number of threads from PARALLEL option applies only to index loading step. The heap scan step is performed by the transaction thread (single thread). During this process, the rows are collected into batches; when a batch is full (reached 16M in size), it is dispatched to a index loading thread (or pushed in the queue), the collecting process continues. Batch size computing consists of index key size and the row identifier size (8 bytes).

    *   The creation of online index is performed in three stages:
    
        * Adding index into schema with status INDEX IS IN ONLINE BUILDING. This is performed with a SCH_M_LOCK (like all schema changes). After this, the lock is demoted to IX_LOCK.

        * Populating the index : scanning the heap file into batches, sorting the batches and adding the keys to index. During this step, other transactions can modify table data (index is also updated with data changes from other committed transactions).

>         * Updating the index status as NORMAL INDEX; this is again performed with a SCH_M_LOCK.

Reviewer: ```suggestion
        * Updating the index status as NORMAL INDEX; this is performed after promoting the table lock back to SCH_M_LOCK (promotion is guaranteed).
```

    *   The session and server timezone (:ref:`timezone-parameters`) should not be changed if database contains indexes or function index on columns of type TIMESTAMP, TIMESTAMP WITH LOCAL TIME ZONE or DATETIME WITH LOCAL TIME ZONE.
    
    *   The leap second support parameter (:ref:`timezone-parameters`) should not be changed if database contains indexes or function index on columns of type TIMESTAMP or TIMESTAMP WITH LOCAL TIME ZONE. 
    
    *   The number of threads from PARALLEL option applies only to index loading step. The heap scan step is performed by the transaction thread (single thread). During this process, the rows are collected into batches; when a batch is full (reached 16M in size), it is dispatched to a index loading thread (or pushed in the queue), the collecting process continues. Batch size computing consists of index key size and the row identifier size (8 bytes).

>     *   When using CSQL in stand-alone mode, the WITH ONLINE option is ignored (normal index is created).

Reviewer 2: ```suggestion
    *   Under stand-alone mode, the WITH ONLINE option is ignored (normal index is created).
```

    *   The creation of online index is performed in three stages:
    
        * Adding index into schema with status INDEX IS IN ONLINE BUILDING. This is performed with a SCH_M_LOCK (like all schema changes) on table. After this, the lock is demoted to IX_LOCK.

        * Populating the index : scanning the heap file into batches, sorting the batches and adding the keys to index. During this step, other transactions can modify table data (index is also updated with data changes from other committed transactions).

        * Updating the index status as NORMAL INDEX; this is performed after promoting the table lock back to SCH_M_LOCK (promotion is guaranteed).
    

>     *   The online index being build is displayed by SHOW statements from other transactions. It is not visible from other transactions in :ref:`db-index` system table due to MVCC snapshot (other transactions can only see committed entries in this table).

Reviewer 2: ```suggestion
    *   The online index being built is displayed by SHOW statements from other transactions. It is not visible from other transactions in :ref:`db-index` system table due to MVCC snapshot (other transactions can only see committed entries in this table).
```

    *   The session and server timezone (:ref:`timezone-parameters`) should not be changed if database contains indexes or function index on columns of type TIMESTAMP, TIMESTAMP WITH LOCAL TIME ZONE or DATETIME WITH LOCAL TIME ZONE.
    
    *   The leap second support parameter (:ref:`timezone-parameters`) should not be changed if database contains indexes or function index on columns of type TIMESTAMP or TIMESTAMP WITH LOCAL TIME ZONE. 
    

>     *   The number of threads from PARALLEL option applies only to index loading step. The heap scan step is performed by the transaction thread (single thread). During this process, the rows are collected into batches; when a batch is full (reached 16M in size), it is dispatched to a index loading thread (or pushed in the queue), the collecting process continues. Batch size computing consists of index key size and the OID size (8 bytes).

Reviewer: I am afraid this paragraph is too technical (for instance OID will surely confuse any reader).

Reviewer: Maybe it is better to offer additional information more relevant to use, like how locking works. It first gets SCH_M lock, then demotes to IX? then it promotes in the end back to SCH_M in both cases successful or failed.

Another information worth mentioning is how unique constraint violation is checked (I think it is reported at the end of loading).

Are there other restrictions or differences to normal loading?

Reviewer 2: We may also add an example how csql shows an index being built:
`INDEX idx1 ON usertable (field5) IN PROGRESS`

one more thing also worth addressing is `db_index` catalog table has `status` column. 
online index will be added to the catalog table once it is completed. it won't be shown during online loading. 
you may cover it as a separate patch.

    
    csql> ;sc tbl

Online index creation
---------------------

You can create the index while still allowing other transactions to insert or update the table.

.. code-block:: sql

>     CREATE TABLE tbl (a int, b int, c int);

Reviewer: The first example is not online index loading.

         't1'                            1  'i_t1'                           1  'i1'                  'A'                             0         NULL  NULL                  'YES'                 'BTREE'               NULL                  NULL                  'NO'
       
       1 row selected. (0.020779 sec) Committed.
       
       1 command(s) successfully processed.
       csql> desc t1;
       
       === <Result of SELECT Command in Line 1> ===
       
         Field                 Type                  Null                  Key                   Default               Extra
       ====================================================================================================================================
         'i1'                  'INTEGER'             'YES'                 'MUL'                 NULL                  ''
         'i2'                  'INTEGER'             'YES'                 ''                    NULL                  ''

>        csql> ;sc t1

Reviewer: ```suggestion
       csql> ;schema t1
```
